### PR TITLE
webkitgtk: fix build on darwin

### LINKED ISSUE #110218: Can't build webkitgtk on darwin
**Describe the bug**
Can't build webkitgtk on darwin

**To Reproduce**
1. Try to build gnome3.webkitgtk package on darwin

**Expected behavior**
Package is successfully built

**Additional context**
```
-- Checking for module 'libtasn1'
--   Found libtasn1, version 4.16.0
-- Found Libtasn1: /nix/store/v7x2c8d4kk59mll0sxs36bb4n25a1ama-libtasn1-4.16.0/lib/libtasn1.dylib
CMake Error at Source/cmake/OptionsGTK.cmake:328 (message):
  Either GLX or EGL is needed for ENABLE_GRAPHICS_CONTEXT_GL.
Call Stack (most recent call first):
  Source/cmake/WebKitCommon.cmake:58 (include)
  CMakeLists.txt:173 (include)


-- Configuring incomplete, errors occurred!
See also "/tmp/nix-build-webkitgtk-2.30.3.drv-0/webkitgtk-2.30.3/build/CMakeFiles/CMakeOutput.log".
See also "/tmp/nix-build-webkitgtk-2.30.3.drv-0/webkitgtk-2.30.3/build/CMakeFiles/CMakeError.log".
builder for '/nix/store/yw733bab2shqx9m2arzm2cm7pxyakxfr-webkitgtk-2.30.3.drv' failed with exit code 1
cannot build derivation '/nix/store/x5hzjb5gm6i4k7mkfzlqfa43kl45q230-applications.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/dvdl92l9lw0c424818gl337433ch8yg8-system-path.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/3y6mdri98phszqvnixs8sf44yz83hb89-darwin-system-21.03pre263904.2fbc36f3d89+darwin4.0000000.drv': 1 dependencies couldn't be built
error: build of '/nix/store/3y6mdri98phszqvnixs8sf44yz83hb89-darwin-system-21.03pre263904.2fbc36f3d89+darwin4.0000000.drv' failed
```

**Notify maintainers**

@hedning @jtojnar @worldofpeace @dasj19 

**Metadata**
- system: `"x86_64-darwin"`
 - host os: `Darwin 20.3.0, macOS 10.16`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.3.10`
 - channels(yevhenshymotiuk): `"darwin, home-manager, nixpkgs-21.03pre263904.2fbc36f3d89"`
 - nixpkgs: `/Users/yevhenshymotiuk/.nix-defexpr/channels/nixpkgs`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: gnome3.webkitgtk
# a list of nixos modules affected by the problem
module: webkitgtk
```

### LINKED ISSUE #184328: libproxy: build failure on darwin
### Steps To Reproduce
1. `nix build nixpkgs#libproxy`

### Build log
```
CMake Error at libproxy/cmake/modules/pacrunner_duktape.cmk:10 (px_check_modules):
  Unknown CMake command "px_check_modules".
Call Stack (most recent call first):
  libproxy/cmake/modules.cmk:21 (include)
  libproxy/CMakeLists.txt:14 (include)


-- Configuring incomplete, errors occurred!
```

See [Hydra](https://hydra.nixos.org/job/nixpkgs/trunk/libproxy.x86_64-darwin).

### Additional context

- Regression of #181917
- Blocks: #157874
- Blocks: porting GSound and GNOME Clocks to macOS
- Notify upstream: libproxy/libproxy#161

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```

### PR DESCRIPTION
**Pull request by wegank**
###### Description of changes

As the title says. Note that the PR is still in draft, since Midori crashes on macOS.

Built on my own machine (aarch64-darwin) and on GitHub Actions (x86_64-darwin), where a full build takes 4 hours.

Patches are applied without `lib.optionals stdenv.isDarwin`, as in #184329.

Fixes #110218

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/webkitgtk/default.nix [hunk 1]
```diff
@@ -93,6 +107,14 @@ stdenv.mkDerivation rec {
       src = ./fdo-backend-path.patch;
       wpebackend_fdo = libwpe-fdo;
     })
+  ] ++ [
+    # https://bugs.webkit.org/show_bug.cgi?id=232934
+    ./fix-build-without-opengl-or-es.patch
+    # https://bugs.webkit.org/show_bug.cgi?id=126433
+    ./fix-conflicting-types-on-darwin.patch
+    # non upstreamed patches
+    ./fix-wtf-errors-on-darwin.patch
```
Comment (veprbl) [inline pkgs/development/libraries/webkitgtk/default.nix:116]: I see, these are mostly the changes from patch below. I can't remember why I didn't originally upstream those. Can we do it now?

File: pkgs/development/libraries/webkitgtk/default.nix [hunk 2]
```diff
@@ -93,6 +107,14 @@ stdenv.mkDerivation rec {
...
+    ./disable-libpas-on-darwin.patch # TODO: regenerate me on 2.37.1
```
Comment (veprbl) [inline pkgs/development/libraries/webkitgtk/default.nix:117]: Looks upstreamable.

**Review by veprbl — COMMENTED** [review]
Looks good.

File: pkgs/development/libraries/webkitgtk/default.nix [hunk 3]
```diff
@@ -93,6 +107,14 @@ stdenv.mkDerivation rec {
       src = ./fdo-backend-path.patch;
       wpebackend_fdo = libwpe-fdo;
     })
+  ] ++ [
+    # https://bugs.webkit.org/show_bug.cgi?id=232934
+    ./fix-build-without-opengl-or-es.patch
```
[the remark this answers is not in the document]
Comment (wegank) [reply pkgs/development/libraries/webkitgtk/default.nix:112]: The patch on Bugzilla is outdated, I rewrote it following the same idea.

File: pkgs/development/libraries/webkitgtk/default.nix (hunk 1 above)
Comment (wegank) [reply pkgs/development/libraries/webkitgtk/default.nix:116]: Patch 126433 was actually merged and then reverted.

I'm afraid I don't have time to upstream them these days. I'm welcome if anyone else does it before me.

File: pkgs/development/libraries/libproxy/default.nix [hunk 4]
```diff
@@ -28,6 +28,10 @@ stdenv.mkDerivation rec {
     hash = "sha256-pqj1LwRdOK2CUu3hYIsogQIXxWzShDuKEbDTbtWkgnQ=";
   };
 
+  patches = [
+    ./fix-build-on-darwin.patch
```
Comment (jtojnar) [inline pkgs/development/libraries/libproxy/default.nix:32]: Would be nice to include more detailed description of the build failure in the commit message so that in the future, we can check if the patch is still relevant.

Comment (wegank) [reply pkgs/development/libraries/libproxy/default.nix:32]: I think I'll send the patch upstream and create a new PR for libproxy in nixpkgs later today, since Midori is crashing on macOS...

File: pkgs/development/libraries/webkitgtk/default.nix [hunk 5]
```diff
@@ -65,6 +67,9 @@
 , systemdSupport ? stdenv.isLinux
 }:
 
+let
+  cmakeBool = x: if x then "ON" else "OFF";
```
Comment (jtojnar) [inline pkgs/development/libraries/webkitgtk/default.nix:71]: Why this change?

Comment (wegank) [reply pkgs/development/libraries/webkitgtk/default.nix:71]: Oh, this was moved just for nix formatting. I will change it back.

File: pkgs/development/libraries/webkitgtk/default.nix [hunk 6]
```diff
@@ -93,6 +108,17 @@ stdenv.mkDerivation rec {
       src = ./fdo-backend-path.patch;
       wpebackend_fdo = libwpe-fdo;
     })
+  ] ++ [
+    # https://bugs.webkit.org/show_bug.cgi?id=232934
```
Comment (jtojnar) [inline pkgs/development/libraries/webkitgtk/default.nix:112]: Please include more detailed comments about what each patch does, otherwise I will remove them in the future when I notice the upstream issue is resolved and it builds on Linux without it:

```suggestion
    # Fix build without OpenGL (important on Darwin)
    # https://bugs.webkit.org/show_bug.cgi?id=232934
```

File: pkgs/development/libraries/webkitgtk/default.nix [hunk 7]
```diff
@@ -93,6 +108,17 @@ stdenv.mkDerivation rec {
...
+    (fetchpatch {
+      url = "https://git.yoctoproject.org/poky/plain/meta/recipes-sato/webkit/webkitgtk/0001-Fix-build-without-opengl-or-es.patch";
+      sha256 = "sha256-9MUDNTda2yxo5knhBKu2PQXNASgTW2Z7H4F0aFt4b0Q=";
+    })
+    # https://bugs.webkit.org/show_bug.cgi?id=126433
+    ./fix-conflicting-types-on-darwin.patch
```
Comment (jtojnar) [inline pkgs/development/libraries/webkitgtk/default.nix:118]:
```suggestion

    # Fix conflicting types on Darwin.
    # https://bugs.webkit.org/show_bug.cgi?id=126433
    # This was reverted and replaced by a different fix in
    # https://github.com/WebKit/WebKit/commit/ec643c6f6dd5c3d4702847cb79965029095ce024
    # but it does not work because of ???
    (fetchpatch {
      url = "https://github.com/WebKit/WebKit/commit/68822eb73f2cdd843dbe6bdd346e7268b279650b.patch";
      sha256 = "";
      excludes = [
        "Source/JavaScriptCore/ChangeLog"
      ];
    })
```

File: pkgs/top-level/all-packages.nix [hunk 8]
```diff
@@ -21650,7 +21650,7 @@ with pkgs;
 
   wcslib = callPackage ../development/libraries/science/astronomy/wcslib { };
 
-  webkitgtk = callPackage ../development/libraries/webkitgtk {
+  webkitgtk = darwin.apple_sdk_11_0.callPackage ../development/libraries/webkitgtk {
```
Comment (jtojnar) [inline pkgs/top-level/all-packages.nix:21653]: This deserves an explanation comment in the code.

File: pkgs/development/libraries/webkitgtk/default.nix [hunk 9]
```diff
@@ -78,6 +83,16 @@ stdenv.mkDerivation rec {
     sha256 = "sha256-1VMvqITJQ9xI8ZEUc91mOrpAejs1yqewS6wUGbQeWQg=";
   };
 
+  # Cocoa headers are not packaged in the tarball
```
Comment (jtojnar) [inline pkgs/development/libraries/webkitgtk/default.nix:86]: Could you open a bug upstream so that they are added? Having to update multiple hashes when updating is quite annoying since we can no longer simply use `update-source-version`. Alternatively, you could add an update script ([example](https://github.com/NixOS/nixpkgs/blob/2839e5c7fad06c731008884cac9325f596363432/pkgs/tools/graphics/gmic/default.nix#L80-L99)).

**Comment by veprbl** [comment]
The build almost finishes for me on x86_64-darwin
https://gist.github.com/veprbl/1b0d480354e7714ff5aa93e596112c39

File: pkgs/top-level/all-packages.nix (hunk 8 above)
Comment (wegank) [reply pkgs/top-level/all-packages.nix:21653]: Done.

File: pkgs/development/libraries/webkitgtk/default.nix (hunk 7 above)
Comment (wegank) [reply pkgs/development/libraries/webkitgtk/default.nix]: Done, with an additional postFetch that changes `#ifdef __APPLE__` to `#if defined(__APPLE__)`.

File: pkgs/development/libraries/webkitgtk/default.nix (hunk 6 above)
Comment (wegank) [reply pkgs/development/libraries/webkitgtk/default.nix:110]: Done.

File: pkgs/development/libraries/libproxy/default.nix (hunk 4 above)
Comment (wegank) [reply pkgs/development/libraries/libproxy/default.nix:32]: Moved to #185853.

File: pkgs/development/libraries/webkitgtk/default.nix (hunk 9 above)
Comment (wegank) [reply pkgs/development/libraries/webkitgtk/default.nix:83]: Opened Bug 243759. Or we could simply fetch the tarball from GitHub, probably.

Comment (wegank) [reply pkgs/development/libraries/webkitgtk/default.nix]: Updated the patch so that Cocoa headers are no longer needed.

File: pkgs/development/libraries/webkitgtk/default.nix [hunk 10]
```diff
@@ -195,30 +223,26 @@ stdenv.mkDerivation rec {
...
-    "-DENABLE_MINIBROWSER=OFF"
+    "-DENABLE_GAMEPAD=OFF" # required, we don't have libmanette on darwin
+    "-DENABLE_JOURNALD_LOG=OFF"
     "-DENABLE_QUARTZ_TARGET=ON"
-    "-DENABLE_VIDEO=ON"
-    "-DENABLE_WEBGL=OFF"
-    "-DENABLE_WEB_AUDIO=OFF"
     "-DENABLE_X11_TARGET=OFF"
-    "-DUSE_APPLE_ICU=OFF"
+    "-DUSE_APPLE_ICU=OFF" # recommended, see 220081
```
Comment (jtojnar) [inline pkgs/development/libraries/webkitgtk/default.nix:230]: Can you use full URLs? It is less annoying that way.

File: pkgs/development/libraries/webkitgtk/default.nix (hunk 1 above)
Comment (wegank) [reply pkgs/development/libraries/webkitgtk/default.nix:118]: The patch is now upstreamed.

**Comment by jtojnar** [comment]
See the OP:

> Note that the PR is still in draft, since Midori crashes on macOS.

**Comment by wegank** [comment]
@Sciencentistguy We wouldn't even be able to upgrade this to 2.38.0, for example [this](https://github.com/WebKit/WebKit/blob/6ce9e6ed5d1ace67854008229aef5db10347dc4f/Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp#L34)

```
#include <gio/gdesktopappinfo.h>
```

is not available on Darwin. And the absence of `webkitgtk` on Darwin doesn't really matter in the presence of `darwin.apple_sdk.frameworks.WebKit`.

**Comment by yqrashawn** [comment]
seems the opengl patch link is broken `https://git.yoctoproject.org/poky/plain/meta/recipes-sato/webkit/webkitgtk/0001-Fix-build-without-opengl-or-es.patch`
